lib/kubernetes-deploy/kubernetes_resource/daemon_set.rb

```diff
@@ -0,0 +1,99 @@
+# frozen_string_literal: true
+module KubernetesDeploy
+  class DaemonSet < KubernetesResource
+    TIMEOUT = 5.minutes
+
+    def sync
+      raw_json, _err, st = kubectl.run("get", type, @name, "--output=json")
+      @found = st.success?
+
+      if @found
+        daemonset_data = JSON.parse(raw_json)
+        @desired_number = daemonset_data["status"]["desiredNumberScheduled"]
+        @rollout_data = daemonset_data["status"]
+          .slice("currentNumberScheduled", "desiredNumberScheduled", "numberReady")
+        @status = @rollout_data.map { |state_replicas, num| "#{num} #{state_replicas.chop.pluralize(num)}" }.join(", ")
+        @pods = find_pods(daemonset_data)
+      else # reset
+        @rollout_data = { "currentNumberScheduled" => 0 }
+        @status = nil
+        @pods = []
+      end
+    end
+
+    def deploy_succeeded?
+      @desired_number == @rollout_data["desiredNumberScheduled"].to_i &&
+      @desired_number == @rollout_data["numberReady"].to_i
+    end
+
+    def deploy_failed?
+      @pods.present? && @pods.all?(&:deploy_failed?)
+    end
+
+    def failure_message
+      @pods.map(&:failure_message).compact.uniq.join("\n")
+    end
+
+    def timeout_message
+      @pods.map(&:timeout_message).compact.uniq.join("\n")
+    end
+
+    def deploy_timed_out?
+      super || @pods.present? && @pods.all?(&:deploy_timed_out?)
+    end
+
+    def exists?
+      @found
+    end
+
+    def fetch_events
+      own_events = super
+      return own_events unless @pods.present?
+      most_useful_pod = @pods.find(&:deploy_failed?) || @pods.find(&:deploy_timed_out?) || @pods.first
+      own_events.merge(most_useful_pod.fetch_events)
+    end
+
+    def fetch_logs
+      container_names.each_with_object({}) do |container_name, container_logs|
+        out, _err, _st = kubectl.run(
+          "logs",
+          id,
+          "--container=#{container_name}",
+          "--since-time=#{@deploy_started.to_datetime.rfc3339}",
+          "--tail=#{LOG_LINE_COUNT}"
+        )
+        container_logs[container_name] = out.split("\n")
+      end
+    end
+
+    private
+
+    def unmanaged?
+      @parent.blank?
+    end
+
+    def container_names
+      @definition["spec"]["template"]["spec"]["containers"].map { |c| c["name"] }
+    end
+
+    def find_pods(ds_data)
+      label_string = ds_data["spec"]["selector"]["matchLabels"].map { |k, v| "#{k}=#{v}" }.join(",")
+      raw_json, _err, st = kubectl.run("get", "pods", "-a", "--output=json", "--selector=#{label_string}")
+      return [] unless st.success?
+
+      all_pods = JSON.parse(raw_json)["items"]
+      all_pods.each_with_object([]) do |pod_data, relevant_pods|
+        next unless pod_data["metadata"]["ownerReferences"].any? { |ref| ref["uid"] == ds_data["metadata"]["uid"] }
```

Comment:
Seemingly this will select all pods belonging to the DaemonSet, both old and updated. Is there a way to select only the ones in the updated generation? It's the intermediate ReplicaSet layer that provides this guarantee for Deployments. If it isn't possible to filter down to the relevant pods, then we'll need to change our logic for success/failure.